packages/location-management/src/components/LocationTree/index.tsx

```diff
@@ -1,110 +1,92 @@
+
+/** helper type, shortened form */
+export type TreeNode = TreeModel.Node<ParsedHierarchyNode>;
+
+interface SelectCallbackInfo {
+  event: 'select';
+  selected: boolean;
+  node: TreeNode;
+  selectedNodes: TreeNode;
+  nativeEvent: MouseEvent;
+}
 
 interface TreeProp {
-  data: ParsedHierarchyNode[];
-  OnItemClick: (item: ParsedHierarchyNode) => void;
+  data: TreeNode[];
+  selectedNode: TreeNode | undefined;
+  onSelect: (item: TreeNode | undefined) => void;
+}
+
+interface ExpandCallbackInfo {
+  node: TreeNode;
+  expanded: boolean;
+  nativeEvent: MouseEvent;
 }
 
 export const Tree: React.FC<TreeProp> = (props: TreeProp) => {
-  const { data, OnItemClick } = props;
-  const dispatch = useDispatch();
+  const { data, onSelect, selectedNode } = props;
   const { t } = useTranslation();
 
   const [expandedKeys, setExpandedKeys] = useState<Key[]>([]);
   const [searchValue, setSearchValue] = useState<string>('');
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  const filterData: ParsedHierarchyNode[] = [];
-  const locationTreeState = useSelector((state) => getLocationTreeState(state));
-
-  /**
-   * Function to to expand tree
-   *
-   * @param {string} value - value to be searched and expanded
-   */
-  function expandTree(value: string) {
-    const expandedKeys = filterData
-      .map((item) =>
-        value.length && item.label.toLocaleLowerCase().indexOf(value.toLowerCase()) > -1
-          ? getParentKey(item.id, item.parent as string, filterData)
-          : null
-      )
-      .filter((item, i, self) => item && self.indexOf(item) === i);
-    setExpandedKeys(expandedKeys as string[]);
-    dispatch(
-      setLocationTreeState({
-        keys: expandedKeys as React.Key[],
-        node: locationTreeState?.node as ParsedHierarchyNode,
-      })
-    );
-  }
+  const [searchExpandedKeys, setSearchExpandedKeys] = useState<Key[]>([]);
 
   useEffect(() => {
-    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-    if (locationTreeState?.node) {
-      const node = getHierarchyNodeFromArray(filterData, locationTreeState.node.id);
-      if (node) {
-        onExpand(locationTreeState.keys);
-        OnItemClick(node);
-      }
+    const updateExpandedKeys = (keys: Key[]) => setExpandedKeys([...expandedKeys, ...keys]);
+    // set expanded and selected keys depending on location tree state and searchValue
+    if (selectedNode) {
+      let parentPath = selectedNode.getPath();
+      parentPath = parentPath.slice(1, parentPath.length);
+      const parentKeys = parentPath.map((node) => node.model.id);
+      updateExpandedKeys(parentKeys);
+    }
+    if (searchValue) {
+      let matchedNodes: TreeNode[] = [];
+      data.forEach((tree: TreeNode) => {
+        const matchedNodesPerTree = tree.all(
+          (node) => node.model.title.toLowerCase().indexOf(searchValue.toLowerCase()) > -1
+        );
+        matchedNodes = [...matchedNodes, ...matchedNodesPerTree];
```

Comment:
do we know if this array is emptied between renders? would it have been better to use map/reduce?

Reply:
The array holds all nodes that were matched during a search, not quite sure which renders you mean, changing the search value will re-render the location tree, the array will not be emptied in that case, and will only be populated with new values.

`would it have been better to use map/reduce`
I do not get the motive behind the consideration here, kindly clarify.